internal/handlers/common/distinct_iterator.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"errors"
+	"sync/atomic"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// DistinctIterator returns an iterator that returns a single document containing
```

Comment:
> If we always return a single document as a result, why do we need to return an iterator at all? 

We do the same for `count_iterator`. Not sure what's the exact architectural decision behind that, but I assume that we aim to wrap iterators in iterators to make every "stage" of query processing more consistent(?)

> (so, each new value is returned when Next() is called)

It would still require to store those distinct values in memory, as they need to be sorted.

Reply:
@AlekSi could you explain the reason, please? It seems that I don't really understand the idea of having iterators for these cases.

Reply:
Unlike filter, count accumulate the values from multiple documents to a single one. Distinct on the other hand _could_ return each distinct values as a document, but this is quite awkward. For simplicity, returning it in single document is easier.

Let's assume we have input to distinct like `[{v: 1}, {v: [1, 23]}, {v:true}, {v: 23}, {v:true}]`

current impl for calling Next() returns like this:
```
Next() -> `{'v': bson.Array[1, 23, true]}
Next() -> iterator.ErrIteratorDone
```
if we want to return each value, we would wrap it in document something like below? Also it may complicate array values such as `[1, 23]` as we need to keep in memory for current index of a document and array currently being worked on. 
```
Next() -> `{'v':1}
Next() -> `{'v': 23}
Next() -> `{'v': true}
Next() -> iterator.ErrIteratorDone
```
At the end, we want to return `bson.Array[1, 23, true]` to wire. :hugs:

Reply:
`CountIterator` is being used in the `$count` stage, and all aggregation pipeline stages return an iterator over documents.

But here, we make an iterator for the `distinct` command, not for some aggregation pipeline stage that should fit a specific interface. I think returning just `*types.Array` makes sense. That issue was about _accepting_ the iterator instead of the slice, not about _returning_ the iterator.